### DISCUSSION TITLE
Trip layer doesn't show with altitude on a map with terrain layer

### DISCUSSION THREAD

**Discussion by maevarun** [visgl/deck.gl#6259] [Q&A]
Hey! I try to render a track with terrain layer.
This is a scientific project on an ultra trail race.
I have an .gpx file and I can retrieve the data with latitude and longitude.

![map_without_terrain](https://user-images.githubusercontent.com/91818923/135756281-4bffca30-fdf0-4ddd-933f-5ca0c2ae3448.png)

But how to recover the altitude to reproduce it on the terrain layer ?

![map_terrain](https://user-images.githubusercontent.com/91818923/135756298-4a82d1f4-7206-4aca-8ce4-de67c0585738.png)

Is there a function like get_altitude ? Or the function get_coordinates is enough ? And in this case, how to do this ?

I would appreciate any help!!

I use Python with Jupyter

The code
```py
df2 = df.copy()

df2["coordinates"] = df2["waypoints"].apply(lambda f: [item["coordinates"] for item in f])
df2["timestamps"] = df2["waypoints"].apply(lambda f: [item["timestamp"] - 946691608 for item in f])

df2.drop(["waypoints"], axis=1, inplace=True)

terrain_layer = pdk.Layer(
"TerrainLayer",
elevation_scale=100,
radius=50,
elevation_decoder=ELEVATION_DECODER,
texture=SURFACE_IMAGE,
elevation_data=TERRAIN_IMAGE
)

track_layer = pdk.Layer(
"TripsLayer",
df2,
get_path="coordinates",
get_timestamps="timestamps",
get_color=[253, 128, 93],
opacity=0.8,
width_min_pixels=5,
rounded=True,
fadeTrail = True,
trail_length = 60000000,
current_time = 5000000,
trailLength = 150
)

view_state = pdk.ViewState(
latitude=-21.115141,
longitude=55.536384,
zoom=10.5,
bearing=140,
pitch=60)

mapSector = pdk.Deck(
layers=[terrain_layer, track_layer],
initial_view_state=view_state)

mapSector.to_html("terrain_trip.html")
```

**Comment by kylebarron** [comment]
Most likely your coordinates only have two dimensional data. If you want to display above terrain, your coordinates need to also have a Z value.

**Reply by maevarun** [reply]
Thanks for your answer! You're right. I added the elevation with the longitude and latitude. The result is not perfect. 
It's my first 3D map.

For the data, it's a .gpx file
[tracedetrail.zip](https://github.com/visgl/deck.gl/files/7279838/tracedetrail.zip)

gpx = gpxpy.parse(open('tracedetrail.gpx'))

ls = []

for point in gpx.tracks[0].segments[0].points:
    elem_point = {}
    d1 = point.time
    elem_point["coordinates"] =[point.longitude, point.latitude, point.elevation]
    elem_point["timestamp"] = int(d1.timestamp())
    elem_point["elevation"] = point.elevation
    ls.append(elem_point)

df = pd.DataFrame([[ls]],
                  columns=['waypoints'])

**Reply by chrisgervang** [reply]
Elevation in deck is in meters - not sure what you've got, but wanted to throw that out here in case the scale is an issue.

Also, if you're noticing the imperfect elevation sometimes is covered by terrain you can try turning off the depthTest on the TripLayer with the prop so that it's "always on top"
```python
parameters={ depthTest: False, depthMask: False}
```
(fyi I'm only guessing what this looks like in python, I've done this in JS.

**Reply by maevarun** [reply]
Thanks! I don't know this parameters. The elevation is in meters in the .gpx file. 

I think the imperfection come from some data elevation missing. So now, I try to find the way to replace the nan values by the next values. 

The map look like 
![map_with_trace](https://user-images.githubusercontent.com/91818923/136389012-a1c831da-fe94-4349-8c8c-1eb910f598f0.png)

For the details 
![map_with_terrain](https://user-images.githubusercontent.com/91818923/136389284-1daf7898-65ab-4b74-b4ff-4ed9985ef0ed.png)

Here a sample of my dataframe with the waypoints : 
liste = [{'coordinates': [55.45843, -21.33877, 3.0], 'timestamp': 946689970},
 {'coordinates': [55.45846, -21.33882, None], 'timestamp': 946689980}]

I try to figure out what is the best way : Maybe I can replace the nan values by a fillna with ffill method or drop the coordinates with the nan value.

**Reply by chrisgervang** [reply]
I think ffill is a good workaround too, or making a function that backfills by computing the grade of the two closest elevation values to a missing one (not sure if there is a lib for that).

You could also check out https://github.com/gboeing/osmnx

They have a way to query elevation tiles, add grades and node elevations. https://github.com/gboeing/osmnx-examples/blob/main/notebooks/12-node-elevations-edge-grades.ipynb

A hacky way to get elevation tifs is to download them from the network tab in the chrome dev tools when they're loading into view of your map..

**Reply by maevarun** [reply]
Thank you for the advice! 

A function to replace a missing elevation by the closest elevation was my first idea.  
But I didn't know I can get the elevation in the chrome dev tools. So I can get the elevation in the case I have only the latitude and longitude for coordinates. 

I'll try to figure it out.
